src/s390x/zipl.rs

```diff
@@ -29,22 +39,159 @@ pub fn chreipl<P: AsRef<Path>>(dev: P) -> Result<()> {
     Ok(())
 }
 
-/// Runs `zipl` based on Ignition and BLS configuration in `boot`.
-pub fn zipl<P: AsRef<Path>>(boot: P) -> Result<()> {
+fn secure_execution_is_enabled() -> Result<bool> {
+    let sysfs_flag = "/sys/firmware/uv/prot_virt_guest";
+    match File::open(sysfs_flag) {
+        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(false),
+        Err(e) => Err(e).with_context(|| format!("reading {}", sysfs_flag)),
+        Ok(mut f) => {
+            let mut buffer = String::new();
+            f.read_to_string(&mut buffer)?;
+            Ok(buffer.trim() == "1")
+        }
+    }
+}
+
+fn find_files<P: AsRef<Path> + ?Sized>(
+    config_path: &P,
+    f: impl Fn(&PathBuf) -> bool,
+) -> Result<Vec<PathBuf>> {
+    let mut entries = read_dir(&config_path)
+        .with_context(|| format!("reading directory {}", config_path.as_ref().display()))?
+        .filter_map(Result::ok)
+        .filter_map(|e| if f(&e.path()) { Some(e.path()) } else { None })
+        .collect::<Vec<_>>();
+    entries.sort();
+    Ok(entries)
+}
+
+fn regenerate_initrd<P: AsRef<Path>>(source: P, files: &[PathBuf]) -> Result<NamedTempFile> {
+    let source = source.as_ref();
+    let dest = Builder::new()
+        .prefix("initrd")
+        .suffix(".img")
+        .append(true)
+        .tempfile()
+        .context("creating cpio for extras")?;
+
+    // copying from original initrd, on s390x SE it's not concatenated
+    std::fs::copy(source, dest.path())
+        .with_context(|| format!("copying {} to {}", source.display(), dest.path().display()))?;
+
+    // new small initrd with files
+    let mut ino = 0;
+    let mut initrd = GzEncoder::new(dest.as_file(), Compression::default());
+    for path in files {
+        ino += 1;
+        let file = File::open(path)?;
+        let meta = file.metadata()?;
+        let name = &path.to_string_lossy();
+        let mode = meta.permissions().mode();
+        let mut input = BufReader::new(file);
+        let entry = NewcBuilder::new(name)
+            .mode(mode as u32)
+            .uid(meta.uid())
+            .gid(meta.gid())
+            .mtime(meta.mtime() as u32)
+            .ino(ino);
+        let size = if mode & 0o40_000 == 0o40_000 {
+            // directory
+            0
+        } else {
+            meta.len()
+        };
+        let mut writer = entry.write(&mut initrd, size as u32);
+        let mut tmp = [0u8; 1024];
+        copy_exactly_n(&mut input, &mut writer, size as u64, &mut tmp)?;
+        writer.finish()?;
+    }
+    // finishing CPIO
+    trailer(&mut initrd)?;
+    initrd.finish()?.flush()?;
+
+    Ok(dest)
+}
+
+fn generate_sdboot<P: AsRef<Path>>(mountpoint: P, boot: P) -> Result<PathBuf> {
     let boot = boot.as_ref();
+    let kernel = Cell::new(String::new());
+    let initrd = Cell::new(String::new());
+    let options = Cell::new(String::new());
```

Comment:
`FnMut` requires some other places to be fixed,  i'm not sure `visit_bls_entry` should be changed just because of single usage on `s390x` . but can do it

Reply:
Here's what I've got working locally:

```diff
diff --git a/src/io/bls.rs b/src/io/bls.rs
index 0b7b1e4..481e1ed 100644
--- a/src/io/bls.rs
+++ b/src/io/bls.rs
@@ -30,7 +30,7 @@ use std::path::{Path, PathBuf};
 /// Returns `true` if BLS content was modified.
 pub fn visit_bls_entry(
     mountpoint: &Path,
-    f: impl Fn(&str) -> Result<Option<String>>,
+    mut f: impl FnMut(&str) -> Result<Option<String>>,
 ) -> Result<bool> {
     // walk /boot/loader/entries/*.conf
     let mut config_path = mountpoint.to_path_buf();
diff --git a/src/s390x/zipl.rs b/src/s390x/zipl.rs
index f812e62..d5b6a8a 100644
--- a/src/s390x/zipl.rs
+++ b/src/s390x/zipl.rs
@@ -23,7 +23,6 @@ use flate2::write::GzEncoder;
 use flate2::Compression;
 use nix::mount::MsFlags;
 use regex::Regex;
-use std::cell::Cell;
 use std::fs::{copy, create_dir_all, read_dir, File};
 use std::io::{BufReader, Read, Write};
 use std::os::unix::prelude::MetadataExt;
@@ -114,30 +113,25 @@ fn regenerate_initrd<P: AsRef<Path>>(source: P, files: &[PathBuf]) -> Result<Nam

 fn generate_sdboot<P: AsRef<Path>>(mountpoint: P, boot: P) -> Result<PathBuf> {
     let boot = boot.as_ref();
-    let kernel = Cell::new(String::new());
-    let initrd = Cell::new(String::new());
-    let options = Cell::new(String::new());
+    let mut kernel: Option<String> = None;
+    let mut initrd: Option<String> = None;
+    let mut options: Option<String> = None;

-    let read_opts = |contents: &str| {
+    visit_bls_entry(boot, |contents: &str| {
         for l in contents.lines() {
-            if let Some(s) = l.strip_prefix("linux ") {
-                kernel.set(s.trim().to_owned());
-            } else if let Some(s) = l.strip_prefix("initrd ") {
-                initrd.set(s.trim().to_owned());
-            } else if let Some(s) = l.strip_prefix("options ") {
-                options.set(s.trim().to_owned());
+            match l.split_once(' ') {
+                Some(("linux", v)) => kernel = Some(v.trim().to_owned()),
+                Some(("initrd", v)) => initrd = Some(v.trim().to_owned()),
+                Some(("options", v)) => options = Some(v.trim().to_owned()),
+                _ => {}
             }
         }
         Ok(None)
-    };
-    visit_bls_entry(boot, read_opts)?;
+    })?;

-    let kernel = kernel.into_inner();
-    let initrd = initrd.into_inner();
-    let options = options.into_inner();
-    if kernel.is_empty() || initrd.is_empty() || options.is_empty() {
-        bail!("bls config doesn't have kernel/initrd/options");
-    }
+    let kernel = kernel.ok_or_else(|| anyhow!("missing 'linux' key in default BLS config"))?;
+    let initrd = initrd.ok_or_else(|| anyhow!("missing 'initrd' key in default BLS config"))?;
+    let options = options.ok_or_else(|| anyhow!("missing 'options' key in default BLS config"))?;
     let kernel = format!("{}{}", boot.to_string_lossy(), kernel);
     let initrd = format!("{}{}", boot.to_string_lossy(), initrd);

```